<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span><ul class="toc-item"><li><span><a href="#Импортирование-библиотек" data-toc-modified-id="Импортирование-библиотек-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Импортирование библиотек</a></span></li><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Изучение-данных" data-toc-modified-id="Изучение-данных-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Изучение данных</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Предобработка" data-toc-modified-id="Предобработка-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка</a></span><ul class="toc-item"><li><span><a href="#Сохранение-стоп-слов" data-toc-modified-id="Сохранение-стоп-слов-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Сохранение стоп слов</a></span></li><li><span><a href="#Лемматезация-и-токенизация" data-toc-modified-id="Лемматезация-и-токенизация-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Лемматезация и токенизация</a></span></li><li><span><a href="#SpaCy" data-toc-modified-id="SpaCy-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>SpaCy</a></span></li><li><span><a href="#Разделение-данных" data-toc-modified-id="Разделение-данных-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Разделение данных</a></span></li><li><span><a href="#Векторизация" data-toc-modified-id="Векторизация-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Векторизация</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Обучение</a></span><ul class="toc-item"><li><span><a href="#Обучение-с-лемматизацией-wordnet" data-toc-modified-id="Обучение-с-лемматизацией-wordnet-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Обучение с лемматизацией wordnet</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Дерево решений</a></span></li></ul></li><li><span><a href="#Обучение-с-лемматизацией--sc" data-toc-modified-id="Обучение-с-лемматизацией--sc-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Обучение с лемматизацией  sc</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Дерево решений</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Обучение-и-прогназирование-на-train-и-test-выборках" data-toc-modified-id="Обучение-и-прогназирование-на-train-и-test-выборках-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Обучение и прогназирование на train и test выборках</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Решающие-деревья" data-toc-modified-id="Решающие-деревья-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Решающие деревья</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Выводы</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Задачи**

1. Загрузить и подготовить данные.
2. Обучить разные модели. 
3. Сформулировать выводы.


## Подготовка

### Импортирование библиотек

In [1]:
import pandas as pd
import nltk
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
import re
from tqdm import notebook
notebook.tqdm.pandas()
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score
from sklearn.tree import DecisionTreeClassifier
import spacy
from nltk.corpus import wordnet
import nltk
nltk.download('averaged_perceptron_tagger')

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /home/jovyan/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.


True

### Загрузка данных


In [2]:
data = pd.read_csv('/datasets/toxic_comments.csv')

### Изучение данных

In [3]:
data.head()

Unnamed: 0                                               text  toxic
0           0  Explanation\nWhy the edits made under my usern...      0
1           1  D'aww! He matches this background colour I'm s...      0
2           2  Hey man, I'm really not trying to edit war. It...      0
3           3  "\nMore\nI can't make any real suggestions on ...      0
4           4  You, sir, are my hero. Any chance you remember...      0

In [4]:
print(data[data['Unnamed: 0'].duplicated()])

Empty DataFrame
Columns: [Unnamed: 0, text, toxic]
Index: []


In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159292 entries, 0 to 159291
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   Unnamed: 0  159292 non-null  int64 
 1   text        159292 non-null  object
 2   toxic       159292 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 3.6+ MB


Столбец "Unnamed: 0" не представляет никакой ценности, т.к просто повторяет индексацию  

In [6]:
data = data.drop('Unnamed: 0',axis = 1)

### Вывод

Данные были успешно загружены и изучены.

Всего в данных 159292 строк

Столбец text содержит текст комментария, toxic - целевой признак

Столбец 'Unnamed: 0' - повторение индексов, поэтому он был удален

## Предобработка

### Сохранение стоп слов

In [7]:
nltk.download('stopwords') 

[nltk_data] Downloading package stopwords to /home/jovyan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [8]:
stop_words = set(stopwords.words('english'))

### Лемматезация и токенизация

In [9]:
lemmatizer = WordNetLemmatizer()

In [10]:
def get_wordnet_pos(word):
    """Map POS tag to first character lemmatize() accepts"""
    tag = nltk.pos_tag([word])[0][1][0].upper()
    tag_dict = {"J": wordnet.ADJ,
                "N": wordnet.NOUN,
                "V": wordnet.VERB,
                "R": wordnet.ADV}
    return tag_dict.get(tag, wordnet.NOUN)

In [11]:
def token_lem(text):
    text = text.lower()
    tokens = ' '.join(word_tokenize(text))
    tokens = re.sub(r"[^a-zA-Z']", ' ', tokens).split()
    # lemmatized_tokens = lemmatizer.lemmatize(tokens)
    lemmatized_tokens = [lemmatizer.lemmatize(w, get_wordnet_pos(w)) for w in tokens]
    return ' '.join(lemmatized_tokens)

In [153]:
data['text_final'] = data['text'].progress_apply(token_lem)

  0%|          | 0/159292 [00:00<?, ?it/s]

### SpaCy

In [13]:
import sys
!{sys.executable} -m pip install spacy

!{sys.executable} -m spacy download en

⚠ As of spaCy v3.0, shortcuts like 'en' are deprecated. Please use the
full pipeline package name 'en_core_web_sm' instead.
     |████████████████████████████████| 13.9 MB 2.6 MB/s eta 0:00:01
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')


In [14]:
import spacy
# Initialize spacy 'en' model, keeping only tagger component needed for lemmatization
nlp = spacy.load('en_core_web_sm', disable=['parser', 'ner'])

In [15]:
def sc(text):
    text = ' '.join(re.sub(r"[^a-zA-Z']", ' ', text).split())
    final = nlp(text)
    return(" ".join([token.lemma_ for token in final]))

In [157]:
data['text_final_sc'] = data['text'].progress_apply(sc)

  0%|          | 0/159292 [00:00<?, ?it/s]

In [16]:
data['text_final_sc'] = data['text'].progress_apply(sc)

  0%|          | 0/5 [00:00<?, ?it/s]

### Разделение данных

In [17]:
display(data['text_final'][0])

"explanation why the edits make under my username hardcore metallica fan be revert they be n't vandalism just closure on some gas after i vote at new york doll fac and please do n't remove the template from the talk page since i 'm retire now"

In [18]:
display(data['text_final_sc'][0])

'explanation why the edit make under my username Hardcore Metallica Fan be revert they be not vandalism just closure on some gas after I vote at New York Dolls FAC and please do not remove the template from the talk page since I be retire now'

In [160]:
features = data['text_final']
target = data['toxic']

In [161]:
features_train, features_test, target_train, target_test = train_test_split(features,
                                                                            target,
                                                                            test_size = 0.2,
                                                                            random_state = 12345,
                                                                            stratify=target)

In [162]:
features_sc = data['text_final_sc']
target_sc = data['toxic']

In [163]:
features_train_sc, features_test_sc, target_train_sc, target_test_sc = train_test_split(features_sc,
                                                                            target_sc,
                                                                            test_size = 0.2,
                                                                            random_state = 12345,
                                                                            stratify=target_sc)

### Векторизация

In [164]:
tf_idf = TfidfVectorizer(stop_words=stop_words)

In [165]:
features_train = tf_idf.fit_transform(features_train)
features_test = tf_idf.transform(features_test)

In [166]:
features_train_sc = tf_idf.fit_transform(features_train_sc)
features_test_sc = tf_idf.transform(features_test_sc)

In [167]:
features_train_v, features_valid, target_train_v, target_valid = train_test_split(features_train,
                                                                            target_train,
                                                                            test_size = 0.15,
                                                                            random_state = 12345,
                                                                            stratify=target_train)

In [168]:
features_train_v_sc, features_valid_sc, target_train_v_sc, target_valid_sc = train_test_split(features_train_sc,
                                                                            target_train_sc,
                                                                            test_size = 0.15,
                                                                            random_state = 12345,
                                                                            stratify=target_train_sc)

In [169]:
print(features_train[:5].toarray())

[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]


### Вывод

Данные были успешно подготовленны к обучению модели по ним. Комментарии были токенизированы, лемматизированны, после чего был создан вектор tf_idf

## Обучение

### Обучение с лемматизацией wordnet

#### Логистическая регрессия

In [170]:
%%time
model = LogisticRegression(random_state=12345, class_weight='balanced',max_iter=500)
model.fit(features_train_v, target_train_v)
predictions = model.predict(features_valid)
    
valid_f1_score = f1_score(target_valid, predictions)

CPU times: user 16.3 s, sys: 15.8 s, total: 32.1 s
Wall time: 32.1 s


In [171]:
print(valid_f1_score)

0.7383496249147534


#### Дерево решений

In [172]:
%%time
b_depth = 0
b_score = 0

for depth in [5,10,15,20,30]:
    model = DecisionTreeClassifier(random_state=12345, max_depth = depth, class_weight = 'balanced')
    model.fit(features_train_v, target_train_v)
    predictions = model.predict(features_valid)
    f1 = f1_score(target_valid, predictions)
    if f1 > b_score:
        b_score = f1
        b_depth = depth
print(f'При глубине {b_depth} удалось получить f1_score = {b_score}')

При глубине 30 удалось получить f1_score = 0.641816143497758
CPU times: user 1min 42s, sys: 71.2 ms, total: 1min 42s
Wall time: 1min 42s


### Обучение с лемматизацией  sc

#### Логистическая регрессия

In [173]:
%%time
model = LogisticRegression(random_state=12345, class_weight='balanced',max_iter=500)
model.fit(features_train_v_sc, target_train_v_sc)
predictions = model.predict(features_valid_sc)
    
valid_f1_score = f1_score(target_valid_sc, predictions)

CPU times: user 27.8 s, sys: 27.6 s, total: 55.4 s
Wall time: 55.4 s


In [174]:
print(valid_f1_score)

0.7489693082913421


#### Дерево решений

In [175]:
%%time
b_depth = 0
b_score = 0

for depth in [5,10,15,20,30]:
    model = DecisionTreeClassifier(random_state=12345, max_depth = depth, class_weight = 'balanced')
    model.fit(features_train_v_sc, target_train_v_sc)
    predictions = model.predict(features_valid_sc)
    f1 = f1_score(target_valid_sc, predictions)
    if f1 > b_score:
        b_score = f1
        b_depth = depth
print(f'При глубине {b_depth} удалось получить f1_score = {b_score}')

При глубине 30 удалось получить f1_score = 0.6465324384787472
CPU times: user 1min 41s, sys: 154 ms, total: 1min 41s
Wall time: 1min 41s


### Вывод

Модели обученные на данных, лематизированных с помощью spacy, показали лучшие результаты, поэтому на тестовой выборке будем проверять именно их

## Обучение и прогназирование на train и test выборках

###  Логистическая регрессия


In [176]:
model = LogisticRegression(random_state=12345, class_weight='balanced',max_iter=500)
model.fit(features_train_sc, target_train_sc)
predictions = model.predict(features_test_sc)
    
test_f1_score = f1_score(target_test_sc, predictions)

In [177]:
print(test_f1_score)

0.7604871447902571


### Решающие деревья

In [178]:
model = DecisionTreeClassifier(random_state=12345, max_depth = b_depth, class_weight = 'balanced')
model.fit(features_train_sc, target_train_sc)
predictions = model.predict(features_test_sc)
f1 = f1_score(target_test_sc, predictions)
print(f'При глубине {b_depth} удалось добиться f1_score = {f1}')

При глубине 30 удалось добиться f1_score = 0.6448345712356516


### Вывод

Модель логистической регрессии показала лучшие результаты f1_score = 0.7604

## Выводы

В наборе данных содержится 159292 записи.

Столбец “text” включает в себя текст комментария, а “toxic” является целевым признаком.

Столбец ‘Unnamed: 0’ дублирует индексы, поэтому он был исключен.

Данные были успешно подготовлены для обучения модели: комментарии были токенизированы и лемматизированы, после чего был создан вектор tf_idf.

Логистическая регрессия показала наилучший результат с f1_score = 0.7604.